### ISSUE TITLE
Rule proposal: prefer-instantiation-expression

### ISSUE THREAD

**Issue by bschlenk** [typescript-eslint/typescript-eslint#7253]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+new+plugin+rule%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#supported-rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] My proposal specifically checks TypeScript syntax, or it proposes a check that requires type information to be accurate.
- [X] My proposal is not a "formatting rule"; meaning it does not just enforce how code is formatted (whitespace, brace placement, etc).
- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Description

This would check for code that is wrapping a function or class only to simplify the generics. It would convert code into equivalent zero runtime cost code, using [typescript instantiation expressions](https://devblogs.microsoft.com/typescript/announcing-typescript-4-7/#instantiation-expressions).

### Fail Cases

```typescript
// Fail, should use instantiation expression
class StrokeAlignSelect extends Select<StrokeAlign> {}
// Fail, should use instantiation expression
class StrokeAlignOption extends Option<StrokeAlign> {}

interface Box<T> {
    value: T;
}

function makeBox<T>(value: T) {
    return { value };
}

// Fail, should use instantiation expression
function makeHammerBox(hammer: Hammer) {
    return makeBox(hammer);
}
```

### Pass Cases

```typescript
const StrokeAlignSelect = Select<StrokeAlign>;
const StrokeAlignOption = Option<StrokeAlign>;

interface Box<T> {
    value: T;
}

function makeBox<T>(value: T) {
    return { value };
}

const makeHammerBox = makeBox<Hammer>;
```

### Additional Info

Prettier 3.0 now supports this syntax. Before, it would format it in strange ways.

**Comment by JoshuaKGoldberg** [comment]
Whoo, what a unique and specific use case...! I can see the benefit here but it seems _very_ niche. Adding any parameters to the function would cause it to no longer match the rule.

Adding the _`evaluating community engagement`_ label on to see if any other folks have this request. In the meantime I'd suggest publishing as a standalone package using the docs in https://typescript-eslint.io/developers/custom-rules.

**Comment by JoshuaKGoldberg** [comment]
It's been 6 months and this issue hasn't been positively mentioned in other contexts. Closing this out as a wontfix on our end.

If anybody comes across this later and wants it: you can always build new rules and/or fork ours! See docs on https://typescript-eslint.io/developers/custom-rules. Cheers!
